### PR TITLE
Optimize Expensive Line of Sight Calls + Remove Stream from Sensor Nearest

### PR DESCRIPTION
**Pull request by Owen1212055**
A better more generic name for the first patch would be appreciated.

Additionally, if you have any comments please speak up, this is my first thing doing anything like this and I would like all the feedback possible! Additionally with the identity hashmap, I am not sure if it is very necessary but I thought it _may_ still be a nice improvement to have? Unsure if the memory usage is worth it, so i'd like to hear.

### CODE REVIEW

**Comment by MiniDigger** [comment]
are we sure the first patch is actually faster? 
without having looked into it, I would argue Stream#min is faster than List#sort, as the first should be O(n) while the latter is o(n*log(n)), right? 
I guess we can replace the call to sort and just do a loop manually to find the smallest element. 

edit: ah there is even a Collections#min method you should use

File: patches/server/0629-Remove-streams-from-SensorNearest.patch [hunk 1]
```diff
@@ -63,6 +63,35 @@ index 66c90013e52170a657b1a5dbdb99748a19fe55e8..ffd83db0a419ab589e89feeddd3fb038
...
++            return Optional.empty();
++        } else {
++            List<LivingEntity> visibleEntityList = visibleEntities.get();
++            visibleEntityList.removeIf((visibleEntity) -> !this.isMatchingEntity(entity, visibleEntity));
++            if (visibleEntityList.isEmpty()) {
++                return Optional.empty();
++            }
++
++            visibleEntityList.sort(Comparator.comparingDouble(entity::distanceToSqr));
++            return Optional.of(visibleEntityList.get(0));
```
Comment (MiniDigger) [inline patches/server/0629-Remove-streams-from-SensorNearest.patch]:
```suggestion
+            return Optional.of(Collections.min(visibleEntityList, Comparator.comparingDouble(entity::distanceToSqr)));
```

**Review by MiniDigger — CHANGES_REQUESTED** [review]
see above and below

File: patches/server/0629-Remove-streams-from-SensorNearest.patch (hunk 1 above)
Comment (Owen1212055) [reply patches/server/0629-Remove-streams-from-SensorNearest.patch]: Thank you, that is perfect! Unsure what I was thinking over here.

**Comment by kennytv** [comment]
Did you actually do tests on the size of the seen/unseen lists warranting a change here? Given they're cleared with every tick. Changing that for a "idk maybe it helps" isn't really something we need

**Comment by Owen1212055** [comment]
> Did you actually do tests on the size of the lists warranting a change here? Given they're cleared with every tick. Changing that for a "idk maybe it helps" isn't really something we need

Yes, specifically this was very noticable when there were lots of entities close together. As entities that were within a pretty large radius would be searched, and if there were lots of entities in the same area this check would get more and more expensive.

**Comment by JRoy** [comment]
> Yes, specifically this was very noticable when there were lots of entities close together. As entities that were within a pretty large radius would be searched, and if there were lots of entities in the same area this check would get more and more expensive.

If this was tested, could you please post the performance metrics you found

**Comment by Owen1212055** [comment]
> > Yes, specifically this was very noticable when there were lots of entities close together. As entities that were within a pretty large radius would be searched, and if there were lots of entities in the same area this check would get more and more expensive.
> 
> If this was tested, could you please post the performance metrics you found

Would you like me to test without any of my changes, or would you just me to just revert the maps? (before and after)

**Comment by JRoy** [comment]
> Would you like me to test without any of my changes, or would you just me to just revert the maps? (before and after)

Both if it isn't too much to ask

**Comment by Owen1212055** [comment]
Please tell me if my testing method is flawed, this is a learning experience for me.

Testing with around ~1k axolotls (according to /paper entity list) all of these entities being close together.
Additionally, there were 10 invulnerable fish that were inside the cube.
image

 All server settings are unchanged, and these metrics are mojang mapped. I applied each change and showed how it changed TPS and sent a screene of a profile recorded via yourkit for 2 minutes.

Current TPS: ~4.57
Here is with none of my changes,
image
image


Current TPS: ~ 6.49
Looking at the removeif call it does a "line of sight" check. Someone probably much better than me could figure out a way to make that better, but instead I noticed the contains call and replacing it with a map yields this.
image

Then, after looking at net.minecraft.world.entity.ai.sensing.AxolotlAttackablesSensor.isMatchingEntity(LivingEntity, LivingEntity)
this method seemed to check if the entity could be attacked AFTER doing the line of sight check. 
Because this box is mostly axolotls and not entities that can be attacked it totally got rid of the overhead caused by the line of sight check.
image
Most of these checks are now made for checking entities that are nearby the axolotl via a line of sight check. 

Current TPS: ~ 8

-- 
Testing stream -> list change by itself
Looking at this a bit further, this is it without streams. It doesn't seem to have any really noticeable difference. So, this change can probably be dropped.
image


So overall, the ones that probably did make a difference was changing the map and moving the booleans around.

**Comment by Xemorr** [comment]
In AxolotlAttackablesSensor, could isClose or isInWaterOrBubble be moved into the if statement to reduce the calls to Sensor#isEntityAttackable further? Both of them seem fairly cheap to me but I haven't investigated too much.

**Comment by Owen1212055** [comment]
Rebased and edited based on Xemorr's comment.
We will now first compare if the entity is close (math), then if they are hostile / huntable target (hashmap call), isInWaterOrBubble (simple boolean OR block lookup) and then isEntityAttackable.

**Comment by Owen1212055** [comment]
And lastly, shoo

**Comment by Xemorr** [comment]
Why was this closed?

**Comment by Owen1212055** [comment]
My understanding of this PR made upkeeping it a bit difficult. My knowledge on this is quite low, so I’m worried that this may not actually provide any benefits.

**Comment by Owen1212055** [comment]
Tested a bit more, I had a change of mind.

**Comment by Owen1212055** [comment]
Implemented by Mojang in 1.18.

Sensing list is now an intset
The axlotl sensor condition check was reordered

**Comment by Xemorr** [comment]
Mojang optimising their own code!?

**Pull request closed without merging**